### DISCUSSION TITLE
Classes for Testing without access to UPC UA server?

### DISCUSSION THREAD

**Discussion by AKuij** [FreeOpcUa/python-opcua#1264] [Q&A]
Hello Community,
I'm am newly working an a project that uses OPC UA and my task is to write Unittests. 
The problem is, that while developing I do not have direct access to the OPC UA server, 
so testing a lot of the responses gets a lot harder.
My question is: Is there a way to unittest without the need for mock? 
I tried it for a week now and find it very complicated where to mock what and how and so on. 
The tutorials and guides are somehow not helping, so maybe I missed the ability to test without mock in the documentation.
I would appreciate any help, or guidance in the right direction. Thank you.

**Comment by AndreasHeine** [comment]
i dont know what you try to test but requests and responses are standardized by opc foundation
https://reference.opcfoundation.org/v104/

for example the browse request/response
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.8.2/

you can look at our tests aswell:
https://github.com/FreeOpcUa/python-opcua/tree/master/tests

**Reply by AKuij** [reply]
Thanks a lot for your quick response! The resources you supplied look very promising, I'll have a look at them. 
Have a nice weekend.

**Comment by swamper123** [comment]
Hi AKuij,

I know only two options in such a case:

1) You can run a test server. Depending what you need, you may want to run a small server like [server-minimal.py ](https://github.com/FreeOpcUa/python-opcua/blob/master/examples/server-minimal.py). The unittest module offers setUp and tearDown methods, where you can create/remove a server. If you use pytest, you can do the same (a bit different from the syntax).
Take a look in our tests: https://github.com/FreeOpcUa/python-opcua/tree/master/tests

2) If a running server is not possible, you may have no choice to mock. OPC UA has a standard formats for many cases, you can check. Maybe make a wireshark trace, to understand the bits and bytes a bit better.

**Reply by AKuij** [reply]
Running a small test server sounds exactly like what I need! Thank you for your quick response. 
I'll have a deep look into it. Have a nice weekend.

**Reply by AndreasHeine** [reply]
you should ask for a xml export of the real servers addressspace! then you can simply import it in you server.

**Reply by AKuij** [reply]
Wow, that's very good to know and probably saves a lot on time! Thanks!
